7.2.0rc1 crashes on startup: Unrecoverable Error

### Description of the bug:


When I update our internal Bazel repo from `7.1.2` to `7.2.0rc1`, I get the following error on startup:

```
2024/05/22 09:47:04 Downloading https://releases.bazel.build/7.2.0/rc1/bazel-7.2.0rc1-linux-x86_64...
2024/05/22 09:47:04 Skipping basic authentication for releases.bazel.build because no credentials found in /home/peter/.netrc
Downloading: 68 MB out of 68 MB (100%) 
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
Computing main repo mapping: 
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'Key{moduleKey=<root>, override=null}' (requested by nodes 'com.google.devtools.build.lib.bazel.bzlmod.BazelModuleResolutionValue$$Lambda/0x00000008004c51e0@f40784e')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.util.NoSuchElementException
        at java.base/java.util.List.getFirst(Unknown Source)
        at com.google.devtools.build.lib.bazel.bzlmod.ModuleFileFunction.evaluateRootModuleFile(ModuleFileFunction.java:443)
        at com.google.devtools.build.lib.bazel.bzlmod.ModuleFileFunction.computeForRootModule(ModuleFileFunction.java:280)
        at com.google.devtools.build.lib.bazel.bzlmod.ModuleFileFunction.compute(ModuleFileFunction.java:141)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
        ... 7 more
```

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


Locally, in this repo, even `bazel info` crashes.

```
# MODULE.bazel
module(
    name = "nx0",
    version = "0.1",
)

http_archive = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

```

```
# .bazelversion
7.2.0rc1
```

### Which operating system are you running Bazel on?


Ubuntu 22.04.4 LTS (jammy) amd64

### What is the output of `bazel info release`?


release 7.2.0rc1

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


_No response_

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.


first bad commit is https://github.com/bazelbuild/bazel/commit/298d3d619777982a23e6a683a13406b6a6f188d0


### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

I happy to, just let me know!

--- a/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java
@@ -643,18 +643,18 @@ public class ModuleFileGlobals {
       throws EvalException {
     ModuleThreadContext context = ModuleThreadContext.fromOrFail(thread, "use_repo_rule()");
     context.setNonModuleCalled();
-    // The builder for the singular "innate" extension of this module.
+    // Find or create the builder for the singular "innate" extension of this module.
+    for (ModuleExtensionUsageBuilder usageBuilder : context.getExtensionUsageBuilders()) {
+      if (usageBuilder.isForExtension("//:MODULE.bazel", ModuleExtensionId.INNATE_EXTENSION_NAME)) {
+        return new RepoRuleProxy(usageBuilder, bzlFile + '%' + ruleName);
+      }
+    }
     ModuleExtensionUsageBuilder newUsageBuilder =
         new ModuleExtensionUsageBuilder(
             context,
             "//:MODULE.bazel",
             ModuleExtensionId.INNATE_EXTENSION_NAME,
             /* isolate= */ false);
-    for (ModuleExtensionUsageBuilder usageBuilder : context.getExtensionUsageBuilders()) {
-      if (usageBuilder.isForExtension("//:MODULE.bazel", ModuleExtensionId.INNATE_EXTENSION_NAME)) {
-        return new RepoRuleProxy(usageBuilder, bzlFile + '%' + ruleName);
-      }
-    }
     context.getExtensionUsageBuilders().add(newUsageBuilder);
     return new RepoRuleProxy(newUsageBuilder, bzlFile + '%' + ruleName);
   }

--- a/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java
@@ -249,6 +249,10 @@ public class ModuleThreadContext {
     // Build module extension usages and the rest of the module.
     var extensionUsages = ImmutableList.<ModuleExtensionUsage>builder();
     for (var extensionUsageBuilder : extensionUsageBuilders) {
+      if (extensionUsageBuilder.proxyBuilders.isEmpty()) {
+        // This can happen for the special extension used for "use_repo_rule" calls.
+        continue;
+      }
       extensionUsages.add(extensionUsageBuilder.buildUsage());
     }
     return module

--- a/src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunctionTest.java
+++ b/src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunctionTest.java
@@ -1539,6 +1539,20 @@ public class ModuleFileFunctionTest extends FoundationTestCase {
   }
 
   @Test
+  public void testNoUsages() throws Exception {
+    scratch.overwriteFile(
+        rootDirectory.getRelative("MODULE.bazel").getPathString(),
+        "module(name='aaa')",
+        "http_archive = use_repo_rule('@bazel_tools//tools/build_defs/repo:http.bzl',"
+            + " 'http_archive')");
+
+    EvaluationResult<RootModuleFileValue> result =
+        evaluator.evaluate(
+            ImmutableList.of(ModuleFileValue.KEY_FOR_ROOT_MODULE), evaluationContext);
+    assertThat(result.hasError()).isFalse();
+  }
+
+  @Test
   public void testInvalidRepoInPatches() throws Exception {
     scratch.overwriteFile(
         rootDirectory.getRelative("MODULE.bazel").getPathString(),

